jwst/datamodels/util.py

```diff
@@ -195,15 +208,30 @@ def open(init=None, memmap=False, **kwargs):
     return model
 
 
-def _class_from_model_type(hdulist):
+def _class_from_model_type(init):
     """
     Get the model type from the primary header, lookup to get class
+
+    Parameter
+    ---------
+    init: AsdfFile or HDUList
+
+    Return
+    ------
+    new_class: str or None
     """
     from . import _defined_models as defined_models
 
-    if hdulist:
-        primary = hdulist[0]
-        model_type = primary.header.get('DATAMODL')
+    if init:
+        if isinstance(init, fits.hdu.hdulist.HDUList):
+            primary = init[0]
+            model_type = primary.header.get('DATAMODL')
+        elif isinstance(init, asdf.AsdfFile):
+            tree = init.tree
+            if 'meta' in tree and 'model_type' in tree['meta']:
+                model_type = tree['meta']['model_type']
+            else:
+                model_type = None
```

Comment:
Not sure why this line is unreachable by the tests.  Odd.

Reply:
Ah, right.  So then maybe that `test_open_asdf_no_datamodel_class` is not doing anything for us.  Maybe better to have a test that writes out a file not using `datamodels`, just `asdf`, then try reading it in.

Reply:
Added a test that tests `datamodels.open` on an ASDF file created using `asdf.write_to` that doesn't have a datamodel in its tree.